### PR TITLE
First proposal to add structure graph to the distribution

### LINKED ISSUE #9654: [FEATURE] Iterate over submodels of a model and draw submodel graph
It would be useful to be able to iterate over the submodel structure of a given bokeh model,
and to view the submodel graph, both a way to better understand how bokeh is put together and to more efficiently manipulate attributes of submodels.

I have a proposed solution that constructs a networkx DiGraph of submodels and displays it in a bokeh graph using the graph rendering tools. 

I showed this in the [discourse showcase](https://discourse.bokeh.org/t/revealing-the-structure-of-a-bokeh-model/4685) and this issue follows up on a discussion there with @bryevdv.  This issue is going to be associated with  a PR very soon.

### PR DESCRIPTION
**Pull request by jeremy9959**
This PR addresses [9654](https://github.com/bokeh/bokeh/issues/9654).

Main changes: 
I put a new file, structure.py into the bokeh/util directory.  It provides two functions,
make_graph and draw_model, which respectively construct a networkx digraph of a model and its submodels and renders that digraph in a bokeh model.

I also added a new subdirectory under examples, called it tools, and put a brief demonstration notebook in that subdirectory.

- [x] issues: fixes #9654 
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by jeremy9959** [comment]
I don't really understand why the test for macos version 3.7 failed but it doesn't seem to have anything to do with anything I did.  On the other hand I am pretty new at this so perhaps I am missing something.

**Comment by mattpap** [comment]
> the test for macos version 3.7 failed (...)

Unrelated, just a sloppy test.

File: bokeh/util/structure.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    node_hover_tool = HoverTool(
+        tooltips=[("id", "@index"), ("model", "@model"), ("in", "@in")]
+    )
+    plot.add_tools(node_hover_tool)
+
+    x, y = zip(*graph_renderer.layout_provider.graph_layout.values())
+    xinterval = max(max(x) - min(x), 200)
+    yinterval = max(max(y) - min(y), 200)
+    plot.x_range = Range1d(
+        start=min(x) - 0.15 * xinterval, end=max(x) + 0.15 * xinterval
```
Comment (mattpap) [inline bokeh/util/structure.py]: I suppose `DataRange1d(range_padding=0.15)` would be a more idiomatic alternative.

File: examples/tools/ModelStructureExample.ipynb [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+      "\u001b[0;32m~/GitHub/bokeh/bokeh/io/webdriver.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, reuse, kind)\u001b[0m\n\u001b[1;32m    113\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkind\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"firefox\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    114\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 115\u001b[0;31m                 raise RuntimeError(\"Neither firefox/geckodriver nor chromium-browser/chromedriver are available on system PATH. \" \\\n\u001b[0m\u001b[1;32m    116\u001b[0m                                    \"You can install the former with 'conda install -c conda-forge firefox geckodriver'.\")\n\u001b[1;32m    117\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
+      "\u001b[0;31mRuntimeError\u001b[0m: Neither firefox/geckodriver nor chromium-browser/chromedriver are available on system PATH. You can install the former with 'conda install -c conda-forge firefox geckodriver'."
+     ]
+    }
+   ],
+>>>>>>> jeremy
```
Comment (mattpap) [inline examples/tools/ModelStructureExample.ipynb]: This and other instances need to be cleaned up, both merge markers and outputs. In general, commited notebooks must not have any outputs.

**Comment by mattpap** [comment]
It would be good to add an example (maybe instead of the notebook), so that we could test this under our image diff framework. Though, you as an external contributor won't be able to see the outcome (at this point), so maybe we can do this after this PR is merged.

**Comment by bryevdv** [comment]
> It would be good to add an example (maybe instead of the notebook)

@mattpap I suggested a notebook because I think most people might be more likely to interact with a tool like this and also because in a notebook we can more easily add some accompanying explanatory text. I do think a non-notebook example for the CI tests is a good idea, though.

**Comment by jeremy9959** [comment]
I stripped out the output from the notebook and created a (trivial) python example file.
I tried to use DataRange1d instead of my Range1d "by hand" but it didn't work so I went back to my original approach.  I'll have to look at that more closely to see what the problem is.

File: examples/tools/ModelStructureExample.ipynb (hunk 2 above)
Comment (jeremy9959) [reply examples/tools/ModelStructureExample.ipynb]: I'm taking care of this.

File: bokeh/util/structure.py (hunk 1 above)
Comment (jeremy9959) [reply bokeh/util/structure.py]: I tried this
```
plot.x_range = DataRange1d(range_padding=.15)
plot.y_range = DataRange1d(range_padding=.15)
```
but it didn't work, so I went back to the Range1d method.  I'll look more at this later.

**Comment by jeremy9959** [comment]
As I noted on the discourse site, this code fails -- for algorithmic reasons -- on complicated models that involve callbacks and selections.  I am going to rethink it.  In the meantime, I will close this PR and reopen it when I have something that works generally.

**Pull request closed without merging**